mlflow/gateway/__init__.py

```diff
@@ -0,0 +1,136 @@
+import logging
+import os
+import pathlib
+import sys
+import time
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.constants import CONF_PATH_ENV_VAR, GATEWAY_SERVER_STATE_FILE
+from mlflow.gateway.utils import (
+    wait_until_server_starts,
+    is_pid_alive,
+    write_server_state,
+    read_server_state,
+    _delete_server_state,
+    kill_parent_and_child_processes,
+)
+from mlflow.protos.databricks_pb2 import BAD_REQUEST
+
+import subprocess
+
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.DEBUG)
+
+
+def _start_server(config_path: str, host, port, server_state_file: str = GATEWAY_SERVER_STATE_FILE):
+    """
+    Starts the server with the given configuration, host and port.
+
+    This function sets the global server_process, gateway_host, and gateway_port variables.
+    It starts the server by creating a new subprocess that runs the server script.
+
+    Args:
+        config_path (str): Path to the configuration file.
+        host (str): The hostname to use for the server.
+        port (int): The port number to use for the server.
+        server_state_file (str): The location on the local file system to store the server state
+
+    Returns:
+        None
+
+    Raises:
+        subprocess.CalledProcessError: If the server subprocess fails to start.
+    """
+    server_state = read_server_state(server_state_file)
+
+    if server_state:
+        raise MlflowException(
+            f"There is a currently running server instance at pid '{server_state['pid']}'. Please "
+            "terminate the running server instance prior to starting a new instance within this "
+            "context.",
+            error_code=BAD_REQUEST,
+        )
```

Comment:
what if I Ctrl+C the server process? Then I won't ever be able to start another server unless I find this state file and `rm` it. Can we check to see if a process with the same PID still exists? Are PIDs unique enough to make that viable?

Ideally, when the server receives SIGTERM / SIGKILL, we should remove the state file.

Reply:
I haven't seen a duplicate pid yet and I must have run these tests with the debugging logic (to show the pids) over 1000 times. They're usually not even remotely close to one another (although, numeric 'closeness' is irrelevant here as it's not indexed) As far as I understand, the more modern Linux OSs that our users are going to be using (basically all of the cloud VMs) use a PID reuse system that auto-increments up to the max PID (32768 or something like that) and then does a wrap-around to assign a PID if it's not already allocated to some other task. 
OSX uses randomization with some crazy reuse logic though to avoid predictable PID reuse to thwart attackers attempting to pre-allocate and intercept critical processes. 

TL&DR I think the validation of existing PID to statefile PID is a smart addition and I'm going to add it :)